seastar/src/config_and_start_seastar.rs

```diff
@@ -0,0 +1,339 @@
+    /// opts.set_description(description);
+    ///
+    /// assert_eq!(opts.get_description(), description);
+    /// ```
+    pub fn set_description(&mut self, description: &str) {
+        set_description(self.opts.pin_mut(), description);
+    }
+
+    /// Sets the `Options`' number of threads.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use seastar::{AppTemplate, Options};
+    ///
+    /// let mut opts = Options::new();
+    /// let smp = 42;
+    /// opts.set_smp(smp);
+    ///
+    /// assert_eq!(opts.get_smp(), smp);
+    /// ```
+    pub fn set_smp(&mut self, smp: u32) {
+        set_smp(self.opts.pin_mut(), smp);
+    }
+}
+
+impl Default for Options {
+    fn default() -> Self {
+        Options::new()
+    }
+}
+
+/// The object through which the contents of a `main` function would be ran in a Seastar app.
+/// Configurable through [`Options`].
+pub struct AppTemplate {
+    app: UniquePtr<app_template>,
+}
+
+impl AppTemplate {
+    /// Creates a Seastar app based on its configuration (`[seastar-rs::AppTemplate`]).
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use seastar::{AppTemplate, Options};
+    ///
+    /// let app = AppTemplate::new_from_options(Options::default());
+    /// ```
+    pub fn new_from_options(mut opts: Options) -> Self {
+        AppTemplate {
+            app: new_app_template_from_options(opts.opts.pin_mut()),
+        }
+    }
+
+    /// Runs an app with a void callback (the output of which is always 0) and program arguments (argv).
+    /// The app is run on a separate thread.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use seastar::{AppTemplate, Options};
+    ///
+    /// let fut = async move { Ok(()) };
+    ///
+    /// let mut app = AppTemplate::default();
+    /// let args = vec!["hello"];
+    ///
+    /// assert_eq!(app.run_void(&args[..], fut), 0);
+    /// ```
+    pub fn run_void<I, Arg>(
+        &mut self,
+        args: I,
+        fut: impl Future<Output = cxx_async::CxxAsyncResult<()>> + 'static,
+    ) -> i32
+    where
+        I: IntoIterator<Item = Arg>,
+        Arg: AsRef<str>,
+    {
+        let args: Vec<Arg> = args.into_iter().collect();
+        let args: Vec<&str> = args.iter().map(|arg| arg.as_ref()).collect();
+        run_void(self.app.pin_mut(), &args, VoidFuture::fallible_local(fut))
+    }
+
+    /// Runs an app with an int (status code) callback and program arguments (argv).
+    /// The app is run on a separate thread.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use seastar::{AppTemplate, Options};
+    ///
+    /// let fut = async move { Ok(42) };
+    ///
+    /// let mut app = AppTemplate::default();
+    /// let args = vec!["hello"];
+    ///
+    /// assert_eq!(app.run_int(&args[..], fut), 42);
+    /// ```
+    pub fn run_int<I, Arg>(
+        &mut self,
+        args: I,
+        fut: impl Future<Output = cxx_async::CxxAsyncResult<i32>> + 'static,
+    ) -> i32
```

Comment:
It would be nicer to have `Result<>` instead of returning 1 on allocation error.

Reply:
Or maybe just panic!() since we're writing rust

Reply:
I think a panic makes sense - it's consistent with how Rust handles allocation errors. I'll probably convert  `args_as_ptr` to Rust anyway, so we will get this behavior automatically.